### PR TITLE
Make `findTargets` support a `this.target` of type `Element`

### LINKED ISSUE #245: assert.dom(Element).exists() throws even if it exists
Even though officially the `dom` method accepts the `Element` type, when using it that way the `exists` method ends up throwing an error, even if the element exists. At root, this is because `exists` calls `findElements`, which is not currently set up to check for `this.target` being anything other than a string.

I have a PR inbound shortly to add support to `findElements` for having Element targets, but wanted to track it here also.

### PR DESCRIPTION
**Pull request by dwilhelmi**
Resolves #245

### CODE REVIEW

File: lib/assertions/exists.ts [hunk 1]
```diff
@@ -3,15 +3,21 @@ export default function exists(options, message) {
     message = options;
     options = undefined;
   }
+  let label;
+  if (typeof this.target === 'string') {
+    label = this.target;
+  } else if (this.target instanceof Element) {
+    label = (this.target as Element).tagName;
+  }
```
Comment (Turbo87) [inline lib/assertions/exists.ts]: you might want to reuse the existing `targetDescription` property for this. but we should make sure to write tests for all the valid inputs (if they don't all exist yet) and check if the assertion messages are correct.

File: lib/assertions.ts [hunk 2]
```diff
@@ -965,16 +965,18 @@ export default class DOMAssertions {
...
    * @throws TypeError will be thrown if target is an unrecognized type
    */
-  private findElements(): HTMLElement[] {
+  private findElements(): Element[] {
     if (this.target === null) {
       return [];
     } else if (typeof this.target === 'string') {
       return toArray(this.rootElement.querySelectorAll(this.target));
+    } else if (this.target instanceof Element) {
+      return [this.target];
```
Comment (Turbo87) [inline lib/assertions.ts]: `findElements()` is also used by the `isVisible()` assertion, I assume we would have to add the `format()` label change there too

**Comment by scalvert** [comment]
Any chance we can dust off this PR? It looks like there may be some straightforward changes to make.

File: lib/assertions/exists.ts (hunk 1 above)
Comment (scalvert) [reply lib/assertions/exists.ts]: I've moved some of this logic into `targetDescription`. Feel free to critique there!

File: lib/assertions.ts (hunk 2 above)
Comment (scalvert) [reply lib/assertions.ts]: I made sure this worked for both `exists` and `visible`.

**Comment by scalvert** [comment]
@Turbo87 in light of this PR, thoughts on changing all references to `Element` instead of `HTMLElement`? It feels like that base class is more accurate to what the code is using.

**Comment by scalvert** [comment]
Once CI passes this is GTG.

**Comment by scalvert** [comment]
I'll follow up this PR with one that renames all references to `HTMLElement` to `Element`, so that we're correctly reflecting the type in the comments.

**Comment by scalvert** [comment]
@Turbo87 do you have an concerns with me merging this. Do you need to get 👀 on it? I feel as if it's mostly additive.

**Comment by alexhancock** [comment]
New to the conversation, but I would like to see this merged as well

File: lib/__tests__/is-visible.ts [hunk 3]
```diff
@@ -51,6 +51,25 @@ describe('assert.dom(...).isVisible()', () => {
     });
   });
 
+  describe('with Element', () => {
+    test('fails for missing element', () => {
+      document.body.innerHTML = '<h1 class="baz">foo</h1>bar';
+
+      const headingElement = document.querySelector('h2');
+      console.log(headingElement);
```
Comment (Turbo87) [inline lib/__tests__/is-visible.ts]: 🤔

File: lib/__tests__/is-visible.ts [hunk 4]
```diff
@@ -51,6 +51,25 @@ describe('assert.dom(...).isVisible()', () => {
...
+      assert.dom(headingElement).isVisible();
```
Comment (Turbo87) [inline lib/__tests__/is-visible.ts]: I guess we could just pass `undefined` directly here instead of doing the `querySelector()` dance above

File: lib/assertions.ts [hunk 5]
```diff
@@ -1063,6 +1065,12 @@ export default class DOMAssertions {
...
    */
   private get targetDescription() {
-    return elementToString(this.target);
+    if (typeof this.target === 'string') {
+      return this.target;
+    } else if (this.target instanceof Element) {
+      return elementToString(this.target);
+    } else {
+      return '<not found>';
+    }
```
Comment (Turbo87) [inline lib/assertions.ts]: why did you adjust `targetDescription()` instead of adding this to `elementToString()`?

File: tests/acceptance/qunit-dom-test.js [hunk 6]
```diff
@@ -34,6 +34,9 @@ module('Acceptance | qunit-dom', function(hooks) {
     assert.dom('#hidden-input').isNotVisible();
     assert.dom('p').isVisible({ count: 2 });
 
+    let titleElement = document.querySelector('#title');
+    assert.dom(titleElement).isVisible();
```
Comment (Turbo87) [inline tests/acceptance/qunit-dom-test.js]: I don't think we need this test here. the tests here are essentially just smoke tests to see if the library is loaded correctly. there is no need to test every single feature here.

File: lib/__tests__/is-visible.ts (hunk 3 above)
Comment (scalvert) [reply lib/__tests__/is-visible.ts]: Oops.

File: lib/__tests__/is-visible.ts (hunk 4 above)
Comment (scalvert) [reply lib/__tests__/is-visible.ts]: We totally could. I guess this was existing?

Comment (scalvert) [reply lib/__tests__/is-visible.ts]: I think passing `null` is more appropriate, as that's what's returned if an element isn't found.

File: tests/acceptance/qunit-dom-test.js (hunk 6 above)
Comment (scalvert) [reply tests/acceptance/qunit-dom-test.js]: Fair. I'll remove, since we cover `isVisible`/`isNotVisible` above adequately.

File: lib/assertions.ts (hunk 5 above)
Comment (scalvert) [reply lib/assertions.ts]: Good question. I can't remember my thought process. It _does_ feel more appropriate to add to `elementToString()`. Let me look at it again and see what I was thinking.

Comment (scalvert) [reply lib/assertions.ts]: Here comes the long winded reply:

This was added to address a particular case that affects both `is-visible` _and_ `exists`. It was explicitly related to fixing `is-visible`'s tests when an element is not found via `document.querySelector`, and therefore returns `null`. This is not present in the case where a `string` selector is passed. Specifically, this test:

https://github.com/simplabs/qunit-dom/blob/29cade3b969949f1d535774b79225653263c6bc4/lib/__tests__/is-visible.ts#L58-L60

that you called out in a comment above _does_ have value. That's because we don't test for a case where you pass an expected element as the target, but that element isn't found in the DOM. 

The `exists` tests only cover the following case:

https://github.com/simplabs/qunit-dom/blob/29cade3b969949f1d535774b79225653263c6bc4/lib/__tests__/exists.ts#L31

so the `is-visible` tests are slightly more expansive.

The reason this affects `exists` and `is-visible` is that both utilize the following pattern in their assertions:

https://github.com/simplabs/qunit-dom/blob/29cade3b969949f1d535774b79225653263c6bc4/lib/assertions/exists.ts#L7

where they try to find all elements that match the target in order to get a count. When the target is `null`, this returns an empty array via `this.findElements`:

https://github.com/simplabs/qunit-dom/blob/29cade3b969949f1d535774b79225653263c6bc4/lib/assertions.ts#L1053-L1054

So, when the code continues within the respective assertions for `exists` and `is-visible`, we try to utilize `this.targetDescription` to format the description. The `targetDescription` getter expects either a `string` or an `Element`, but in the above case the target is neither; it's `null`. Since it's not an `Element`, it made sense to add this as a fallback here, inside `targetDescription`, so that we can cover any case where `this.targetDescription` is used on a target that's neither a `string` or `Element`. I feel like this is a good fallback. 

The only thing I'd add to this PR at this point is a test to ensure that `exists` does the correct thing when `null` is passed.
